scripts/xdist/find_order_dependent_test_failures.py

```diff
@@ -0,0 +1,221 @@
+    """
+    Runs tests and returns True if failures are found.
+    """
+    global verbose_option
+    pytest_command = _get_pytest_command(output_file_name)
+    test_output = os.popen(pytest_command).read()
+    if verbose_option:
+        print(test_output)
+    failures_search = re.search(r'=== (\d+) failed', test_output)
+    return bool(failures_search) and int(failures_search.group(1)) > 0
+
+
+def _create_and_check_test_files_for_failures(test_list, test_type):
+    """
+    Run the test list to see if there are any failures.
+
+    Keeps around any test files that produced a failure, and deletes
+    the passing files.
+
+    Returns the pytest command to run if failures are found.
+    """
+    print("Testing {}, includes {} test(s)...".format(test_type, len(test_list)))
+    global test_suite_option
+    output_file_name = "{}/{}_failing_test_list_{}_{}.txt".format(
+        OUTPUT_FOLDER_NAME, test_suite_option, test_type, len(test_list)
+    )
+    dirname, basename = os.path.split(output_file_name)
+    # Note: We don't really need a temporary file, and could just output the tests directly
+    # to the command line, but this keeps the verbose  output cleaner.
+    temp_file = tempfile.NamedTemporaryFile(prefix=basename, dir=dirname, delete=False)
+
+    with io.open(temp_file.name, 'w') as output_file:
+        for line in test_list:
+            output_file.write(line + "\n")
+    temp_file.close()
+
+    if _run_tests_and_check_for_failures(temp_file.name):
+        os.rename(temp_file.name, output_file_name)
+        print('- test failures found.')
+        return _get_pytest_command(output_file_name)
+
+    os.remove(temp_file.name)
+    print('- no failures found.')
+    return None
+
+
+def _find_fewest_tests_with_failures(test_list, test_type):
+    """
+    Recursively tests half the tests, finding the smallest number of tests to obtain a failure.
+
+    Returns:
+        (test_list, pytest_command): Tuple with the smallest test_list and the pytest_command
+            to be used for testing. Returns (None, None) if no failures are found.
+    """
+    if len(test_list) <= 1:
+        return None, None
+
+    pytest_command = _create_and_check_test_files_for_failures(test_list, test_type)
+    if not pytest_command:
+        return None, None
+
+    if len(test_list) == 2:
+        return test_list, pytest_command
+
+    half_tests_num = round((len(test_list) - 1) / 2)
+    failing_test = test_list[-1:]
+    test_list_a = test_list[0:half_tests_num] + failing_test
+    test_list_b = test_list[half_tests_num:]
+    failing_test_list_a, pytest_command_a = _find_fewest_tests_with_failures(test_list_a, 'GROUP-A')
+    if failing_test_list_a:
+        return failing_test_list_a, pytest_command_a
+
+    failing_test_list_b, pytest_command_b = _find_fewest_tests_with_failures(test_list_b, 'GROUP-B')
+    if failing_test_list_b:
+        return failing_test_list_b, pytest_command_b
+
+    return test_list, pytest_command
```

Comment:
This would be the case where there's a complex failure scenario where some list of tests fails, but neither half fails by itself? (Each half containing a necessary ingredient for causing failure in the final test.)

Reply:
Yes.  Added a comment in case this wasn't clear.